File: packages/engine/lib/src/planner/assignment.dart
import 'package:meta/meta.dart';

import '../models/enums.dart';
import 'assignment_config.dart';
import 'scoring.dart';

/// One day of the period, with everything the caller already resolved
/// from `Household` and the calendar date.
@immutable
class PeriodDay {
  const PeriodDay({
    required this.date,
    required this.isTrainingDay,
    required this.timeBudgetMinutes,
  });

  final DateTime date;
  final bool isTrainingDay;
  final int timeBudgetMinutes;
}

/// One day's outcome of assignment. `activityId == null` means
/// deliberately empty — a valid outcome, not a gap.
@immutable
class DayAssignment {
  const DayAssignment({required this.date, this.activityId});

  final DateTime date;
  final String? activityId;
}

/// Walks the scored, sorted pool day by day — see
/// `docs/specs/zuweisen.md`. Does not attach a `Reason` (that is step 8)
/// or cross-check the result (that is step 7).
List<DayAssignment> assignToDays({
  required List<PeriodDay> days,
  required List<ScoredActivity> pool,
  required AssignmentConfig config,
}) {
  final assignableCap = _min(
    config.maxActiveSlots,
    _flooredAtZero(days.length - config.minEmptySlots),
  );

  final budgets = days.map((d) => d.timeBudgetMinutes);
  final shortestBudget = budgets.isEmpty ? 0 : budgets.reduce(_min);
  final longestBudget =
      budgets.isEmpty ? 0 : budgets.reduce((a, b) => a > b ? a : b);
  final hasShortDay = shortestBudget < longestBudget;

  final usedActivityIds = <String>{};
  final assignments = <DayAssignment>[];
  var activeCount = 0;
  var trainingCount = 0;
  var previousArousal = 0;

  for (final day in days) {
    ScoredActivity? chosen;

    if (activeCount < assignableCap) {
      for (final candidate in pool) {
        final activity = candidate.activity;
        if (usedActivityIds.contains(activity.id)) continue;

        if (activity.type == ActivityType.training) {
          if (!day.isTrainingDay) continue;
          if (trainingCount >= config.maxTrainingSlots) continue;
        }

        if (activity.durationMin > day.timeBudgetMinutes) continue;

        final isDemanding = activity.arousal >= config.heavyArousalThreshold;
        if (isDemanding &&
            hasShortDay &&
            day.timeBudgetMinutes == shortestBudget) {
          continue;
        }

        if (previousArousal >= config.heavyArousalThreshold &&
            activity.type != ActivityType.rest &&
            activity.type != ActivityType.enrichment) {
          continue;
        }

        if (previousArousal >= config.maxArousalThreshold &&
            activity.arousal >= config.maxArousalThreshold) {
          continue;
        }

        chosen = candidate;
        break;
      }
    }

    if (chosen != null) {
      usedActivityIds.add(chosen.activity.id);
      activeCount++;
      if (chosen.activity.type == ActivityType.training) trainingCount++;
      previousArousal = chosen.activity.arousal;
      assignments
          .add(DayAssignment(date: day.date, activityId: chosen.activity.id));
    } else {
      previousArousal = 0;
      assignments.add(DayAssignment(date: day.date));
    }
  }

  return assignments;
}

int _flooredAtZero(int value) => value < 0 ? 0 : value;

int _min(int a, int b) => a < b ? a : b;

File: infra/supabase/functions/_shared/planner/steps/assignment.ts
import type { ScoredActivity } from './scoring.ts';
import type { AssignmentConfig } from './assignment_config.ts';

/**
 * One day of the period, with everything the caller already resolved
 * from `Household` and the calendar date.
 */
export interface PeriodDay {
  readonly date: Date;
  readonly isTrainingDay: boolean;
  readonly timeBudgetMinutes: number;
}

/**
 * One day's outcome of assignment. `activityId === null` means
 * deliberately empty — a valid outcome, not a gap.
 */
export interface DayAssignment {
  readonly date: Date;
  readonly activityId: string | null;
}

/**
 * Walks the scored, sorted pool day by day — see
 * `docs/specs/zuweisen.md`. Does not attach a `Reason` (that is step 8)
 * or cross-check the result (that is step 7).
 */
export function assignToDays(args: {
  days: readonly PeriodDay[];
  pool: readonly ScoredActivity[];
  config: AssignmentConfig;
}): DayAssignment[] {
  const { days, pool, config } = args;

  const assignableCap = Math.min(
    config.maxActiveSlots,
    flooredAtZero(days.length - config.minEmptySlots),
  );

  const budgets = days.map((d) => d.timeBudgetMinutes);
  const shortestBudget = budgets.length === 0 ? 0 : Math.min(...budgets);
  const longestBudget = budgets.length === 0 ? 0 : Math.max(...budgets);
  const hasShortDay = shortestBudget < longestBudget;

  const usedActivityIds = new Set<string>();
  const assignments: DayAssignment[] = [];
  let activeCount = 0;
  let trainingCount = 0;
  let previousArousal = 0;

  const findCandidate = (day: PeriodDay, dayIndex: number, excludeRest: boolean) => {
    for (const candidate of pool) {
      const activity = candidate.activity;
      if (usedActivityIds.has(activity.id)) continue;

      // Day 1 of the period is never a deliberately quiet day — a brand
      // new plan (or any period's opening day) needs to visibly do
      // something, not open with "rest" (docs/specs/zuweisen.md, "Tag 1
      // ist nie ein Ruhetag"). Only a preference, not absolute: the
      // caller retries without it if nothing else fits at all, so a
      // genuinely empty day 1 never wins over a rest day that does fit.
      if (dayIndex === 0 && excludeRest && activity.type === 'rest') continue;

      if (activity.type === 'training') {
        if (!day.isTrainingDay) continue;
        if (trainingCount >= config.maxTrainingSlots) continue;
      }

      if (activity.durationMin > day.timeBudgetMinutes) continue;

      const isDemanding = activity.arousal >= config.heavyArousalThreshold;
      if (isDemanding && hasShortDay && day.timeBudgetMinutes === shortestBudget) {
        continue;
      }

      if (
        previousArousal >= config.heavyArousalThreshold &&
        activity.type !== 'rest' &&
        activity.type !== 'enrichment'
      ) {
        continue;
      }

      if (
        previousArousal >= config.maxArousalThreshold &&
        activity.arousal >= config.maxArousalThreshold
      ) {
        continue;
      }

      return candidate;
    }
    return undefined;
  };

  for (let dayIndex = 0; dayIndex < days.length; dayIndex++) {
    const day = days[dayIndex]!;
    let chosen: ScoredActivity | undefined;

    if (activeCount < assignableCap) {
      chosen = findCandidate(day, dayIndex, true);
      if (chosen === undefined && dayIndex === 0) {
        chosen = findCandidate(day, dayIndex, false);
      }
    }

    if (chosen !== undefined) {
      usedActivityIds.add(chosen.activity.id);
      activeCount++;
      if (chosen.activity.type === 'training') trainingCount++;
      previousArousal = chosen.activity.arousal;
      assignments.push({ date: day.date, activityId: chosen.activity.id });
    } else {
      previousArousal = 0;
      assignments.push({ date: day.date, activityId: null });
    }
  }

  return assignments;
}

function flooredAtZero(value: number): number {
  return value < 0 ? 0 : value;
}
